Add SELU activation function

fmassa opened the pull request:
Fixes https://github.com/pytorch/pytorch/issues/1768

Also fixes a problem with RReLU in inplace mode.

fmassa committed:
Add SELU activation function

fmassa committed:
Remove unnecessary case

fmassa committed:
Add Function for SELU + tests and fix RReLU inplace

fmassa committed:
Fix extra line in doc

apaszke commented on torch/nn/_functions/thnn/activation.py:
```diff
@@ -95,7 +98,47 @@ def backward(self, grad_output):
+    def __init__(self, inplace=False):
+        super(SELU, self).__init__(inplace)
+        self.alpha = 1.6732632423543772848170429916717
+        self.scale = 1.0507009873554804934193349852946
```
These can be class attrs. Also, can you write it as a new-style Function?

fmassa replied:
I'm not sure we want to allow the user to specify them. The reason why they chose these parameters is to enforce that the activations have on average zero mean and unit std.
I'll modify it to be new-style

apaszke replied:
I didn't mean that they can be user-specified, but that they don't need to be set in `__init__`. You can put the `alpha = ...` in the class body, and it makes `__init__` completely unnecessary in this case

fmassa replied:
Ah ok, I misunderstood your statement. I'm fixing it

apaszke commented on torch/nn/modules/activation.py:
```diff
@@ -245,6 +245,40 @@ def __repr__(self):
+    More details can be found in the paper `Self-Normalizing Neural Networks`_ .
+
+    Args:
+        inplace: can optionally do the operation in-place
```
`inplace (bool, optional): ...`

apaszke reviewed (commented):
All contbuilds are failing

fmassa commented:
Yeah, I just saw that the tests are failing because there is no `legacy` equivalent for SELU. Should we add it or just skip on SELU?

apaszke commented:
No, you only have to move the test spec from `common_nn.py` to `test_nn.py`

fmassa committed:
Fix tests

Remove in-place tests for RReLU. For some reason they fail on legacy nn, but passes on nn

fmassa committed:
SELU in new-style Function

It also supports double backprop, verifyed with gradgradcheck

fmassa commented:
Added new-style Function to `SELU`. Double backprop was verified with `gradgradcheck`.

fmassa committed:
Fix flake8

andreh7 commented on torch/nn/_functions/thnn/activation.py:
```diff
@@ -95,11 +98,57 @@ def backward(self, grad_output):
 
+class SELU(InplaceFunction):
+    alpha = 1.6732632423543772848170429916717
+    scale = 1.0507009873554804934193349852946
```
would it make sense to add a comment that these two values are fixed point solutions to equations (4) and (5) in the paper (arxiv:1706.02515) for input/output mean zero and unit variance and a normalized weight vector and that their analytical expression is given in equation (14) ?
